service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java

```diff
@@ -185,9 +202,40 @@ public PipelineRun getPipelineRun(UUID jobId, String userId) {
    * status column here. It is currently possible to mark an incomplete pipeline_run as is_success =
    * True using this method.
    */
-  public PipelineRun markPipelineRunSuccess(UUID jobId, String userId) {
+  public PipelineRun markPipelineRunSuccessAndWriteOutputs(
+      UUID jobId, String userId, Map<String, String> outputs) {
     PipelineRun pipelineRun = getPipelineRun(jobId, userId);
     pipelineRun.setIsSuccess(true);
+    pipelineRun.setOutput(outputs);
+
     return pipelineRunsRepository.save(pipelineRun);
   }
+
+  /**
+   * Extract the pipeline outputs from a pipelineRun object, fetch SAS tokens for (currently all of)
+   * them, and return an ApiPipelineRunOutput object with the formatted outputs.
+   *
+   * @param pipelineRun object from the pipelineRunsRepository
+   * @return ApiPipelineRunOutput
+   */
+  public ApiPipelineRunOutput formatPipelineRunOutputs(
+      PipelineRun pipelineRun, String accessToken) {
+    Map<String, String> rawOutputMap = pipelineRun.getOutput();
+
+    // currently all outputs are paths that will need a SAS token
+    Map<String, String> formattedOutputs =
+        rawOutputMap.entrySet().stream()
```

Comment:
ugh this feels kind of gross having to make so many calls to wsm.  also it feels weird to have a `.stream()` followed by a `.collect`.  Is that just standard practice for this type of thing?

Reply:
thanks, yes, this can be cleaner/simpler.